File: scripts/refresh_probe.py
# SPDX-License-Identifier: Apache-2.0
"""Synthetic process-restart host; no real provider or customer data."""
import json
import sys
from types import SimpleNamespace
from psp_cdl_api_server.persistence import WorkflowStore, OwnerCoordinator
from psp_cdl_api_server.sqlite import SqliteBackend
from psp_cdl_core.crypto import sign_envelope
from psp_cdl_llmproxy import RefreshingLlmLoop, prompt_context
from psp_cdl_mcpproxy import McpDispatchGate, binding_digest

sys.stdin.reconfigure(encoding="utf-8");sys.stdout.reconfigure(encoding="utf-8")
options=json.load(sys.stdin)
requests,refreshes,events=[],[],[]
key=bytes([19])*32
def sign(b,version,timestamp,expires):
    return sign_envelope("System "+version,{"algorithm":"hmac-sha256","signatureVersion":"2.0","secretId":"test-signing-key","timestamp":timestamp,"expires":expires,"version":version,"sectionType":"system","contentType":"text","attributes":{**prompt_context(b),"refresh-policy":"interval|expiration","refresh-interval":"1","refresh-grace":"10"}},key)
class Host:
    def authenticate(self,t):return {**options["actor"],"scopes":["sessions:read","sessions:write","models:invoke","tools:list","tools:call"]} if t=="synthetic" else None
    def now(self):return options["now"]
    def snapshot(self,p,s):return {"revision":"a1","policyVersion":s["policyVersion"],"registryRevision":"r1","providerId":"mock","providerRevision":"model-1","expires":190,"releaseSources":[],"releaseComplete":True}
    def prompt(self,p,b):return sign(b,b.get("refresh",{}).get("current_version") or "1.0.0",90,150)
    def verification(self,*_):return {"keys":[{"id":"test-signing-key","algorithm":"hmac-sha256","material":key,"status":"active","trustLevels":[2],"sectionTypes":["system"],"scope":{},"allowUnscoped":False}]}
    def policy(self,p,b,*_):return {"bindingDigest":binding_digest(b),"resources":[{"classes":[],"covenants":["no-training"],"capabilities":[],"checks":{},"parameters":{},"context":{}}]}
    def authorize_final(self,*_):return True
    def plan_turn(self,*_):return {"state":{"done":False},"retained":{"covenants":["no-training"]},"complete":False}
    def authorize_transition(self,*_):return True
    def audit(self,*_):return True
    def authorize_recovery(self,*_):return True
    recovery_policy=policy
    def refresh(self,p,b,r):
        refreshes.append(r)
        return sign(b,options.get("nextVersion","1.0.1"),options["now"],options["now"]+50)
    def authorize_refresh(self,*_):return True
    def audit_refresh(self,p,e):events.append(e);return True
def invoke(r,_):requests.append(r);return {"type":"final","text":"Finished 🧪"}
host=Host();backend=SqliteBackend(sys.argv[1],"peer-epoch",host.now)
try:
    store=WorkflowStore(backend,resume_secret=bytes([42])*32,authorize_persistence=lambda *_:True,coordinator=OwnerCoordinator(),durable_turns=True,prompt_refresh=True)
    gate_host=SimpleNamespace(authenticate=host.authenticate,now=host.now,policy=host.policy,snapshot=lambda p,s:{k:v for k,v in host.snapshot(p,s).items() if k not in ("providerId","providerRevision")})
    gate=McpDispatchGate(store,gate_host,"r1",[])
    loop=RefreshingLlmLoop(store,gate,host,{"id":"mock","revision":"model-1","complete":True,"sources":[],"invoke":invoke},{"postCompletion":"lockdown"})
    controls={"deadline":190,"cancelled":lambda:False,"maxSteps":1,"requestId":options["requestId"],"expectedVersion":options["expectedVersion"]}
    result=loop.recover("synthetic",options["sessionId"],options["requestId"],controls) if options.get("recover") else loop.run("synthetic",options["sessionId"],{"message":"Continue."},controls)
    prompt=store.execute(options["actor"],{"action":"getPromptState","sessionId":options["sessionId"]})
    print(json.dumps({"result":result,"prompt":prompt,"requests":requests,"refreshes":refreshes,"events":events},ensure_ascii=False))
finally:backend.close()

File: scripts/refresh-probe.mjs
// SPDX-License-Identifier: Apache-2.0
// Synthetic process-restart host; no real provider, credential or customer data.
import {readFileSync} from 'node:fs';
import {WorkflowStore,OwnerCoordinator} from '@psp-cdl/api-server/persistence';
import {SqliteBackend} from '@psp-cdl/api-server/sqlite';
import {signEnvelope} from '@psp-cdl/core/crypto';
import {RefreshingLlmLoop,promptContext} from '@psp-cdl/llmproxy';
import {McpDispatchGate,bindingDigest} from '@psp-cdl/mcpproxy';
const options=JSON.parse(readFileSync(0,'utf8')),requests=[],refreshes=[],events=[];
const key=new Uint8Array(32).fill(19),sign=(b,version,timestamp,expires)=>signEnvelope('System '+version,{
  algorithm:'hmac-sha256',signatureVersion:'2.0',secretId:'test-signing-key',timestamp,expires,version,sectionType:'system',contentType:'text',
  attributes:{...promptContext(b),'refresh-policy':'interval|expiration','refresh-interval':'1','refresh-grace':'10'}
},key);
const host={authenticate:t=>t==='synthetic'?{...options.actor,scopes:['sessions:read','sessions:write','models:invoke','tools:list','tools:call']}:null,now:()=>options.now,
  snapshot:(_p,s)=>({revision:'a1',policyVersion:s.policyVersion,registryRevision:'r1',providerId:'mock',providerRevision:'model-1',expires:190,releaseSources:[],releaseComplete:true}),
  prompt:(_p,b)=>sign(b,b.refresh?.current_version??'1.0.0',90,150),
  verification:()=>({keys:[{id:'test-signing-key',algorithm:'hmac-sha256',material:key,status:'active',trustLevels:[2],sectionTypes:['system'],scope:{},allowUnscoped:false}]}),
  policy:(_p,b)=>({bindingDigest:bindingDigest(b),resources:[{classes:[],covenants:['no-training'],capabilities:[],checks:{},parameters:{},context:{}}]}),
  authorizeFinal:()=>true,planTurn:()=>({state:{done:false},retained:{covenants:['no-training']},complete:false}),authorizeTransition:()=>true,
  audit:()=>true,authorizeRecovery:()=>true,
  refresh:(_p,b,r)=>{refreshes.push(r);return sign(b,options.nextVersion??'1.0.1',options.now,options.now+50);},
  authorizeRefresh:()=>true,auditRefresh:(_p,e)=>{events.push(e);return true;}
};
host.recoveryPolicy=host.policy;
const backend=new SqliteBackend(process.argv[2],'peer-epoch',host.now);
try {
  const store=new WorkflowStore(backend,{resumeSecret:new Uint8Array(32).fill(42),authorizePersistence:()=>true,coordinator:new OwnerCoordinator(),durableTurns:true,promptRefresh:true});
  const gateHost={...host,snapshot:(p,s)=>{const {providerId,providerRevision,...a}=host.snapshot(p,s);return a;}};
  const gate=new McpDispatchGate(store,gateHost,'r1',[]);
  const loop=new RefreshingLlmLoop(store,gate,host,{id:'mock',revision:'model-1',complete:true,sources:[],invoke:r=>{requests.push(r);return {type:'final',text:'Finished 🧪'};}},{postCompletion:'lockdown'});
  const controls={deadline:190,cancelled:()=>false,maxSteps:1,requestId:options.requestId,expectedVersion:options.expectedVersion};
  const result=options.recover?await loop.recover('synthetic',options.sessionId,options.requestId,controls):await loop.run('synthetic',options.sessionId,{message:'Continue.'},controls);
  const prompt=await store.execute(options.actor,{action:'getPromptState',sessionId:options.sessionId});
  console.log(JSON.stringify({result,prompt,requests,refreshes,events}));
}finally{backend.close();}
